### PR TITLE
Zauber Cauldrons 1.17

### PR DESCRIPTION
**Pull request by Bloo-dev**
### Feature Changes
This update brings some new features to Zauber Cauldrons, mainly allowing Zauber Cauldrons to use powder snow as a liquid.
Powder snow provides additional cooling, making some more exothermic recipes safe! This means:
- [x] Powder snow makes the use of splash or lingering potions in tier IV potion recipes safe. This allows you to craft tier IV splash and lingering potions
- [x] Powder snow prevents tier IV strength potions from causing fire

In addition to that, there are also some minor changes to features such as:

- [x] Crafting a tier IV poison potion now uses a better source of randomness for the placement of cave spiders and cobwebs
- [x] Crafting a tier IV strength potion now uses a better source of randomness for the placement of fire
- [x] Tier IV potion side effects were tweaked slightly
- [x] Balance changes to Zauber Armor's attribute values
- [x] Zauber Armor flavor speed new uses sugar instead of rabbit's foot
- [x] Zauber Armor flavor knockback resistance was is now armor boost, using turtle shells as an ingredient
- [x] Add Zauber Armor set bonuses
- [x] Zauber Crystal crafting is now based on tier IV potions instead of protection books
- [x] Changes to Zauber Crystal texture
- [x] Magic in a bottle take longer to break and telegraph the breaking before they do
- [x] Campfires can be used to heat Zauber Cauldrons

### Technical Changes
This update completely reworks the Zauber Cauldrons' crafting system in order to make it loot table based. This will make it easier for other modules to interact with Zauber Cauldrons and allow for easier debugging // creative play.

- [x] Rework Zauber Potions crafting
- [x] Rework Zauber Armor crafting
- [x] Rework Zauber Crystals crafting
- [x] Rework Prismarine Shard crafting
- [x] Rework Flower Luck crafting
- [x] Rework Wormhole crafting
- [x] Rework Multi-Sip crafting

In addition to the changes to the crafting system, some parameters for crafting are now dynamic, allowing for more cross-module interactions - especially for liquid tanks.

- [x] Allow arbitrary amounts of sips per bottle
- [x] Allow remote execution of recipes, both with soul and normal fire

Efficiency of Zauber Cauldrons has also improved, especially in worlds with lots of entities. The new crafting system no longer uses `@e` selectors to detect ingredients; instead they refer to a data storage. Zauber Cauldrons' fake players are now prefix according to Gamemode 4's style guide.

- [x] Transition to `marker`-type entities

Due to a change in how forceload works, wormholes into unloaded chunks may not work as expected. Wormholes require extra-thorough testing.

### The features ticked-off above are ready for review. Unchecked features remain to be implemented.

### CODE REVIEW

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/instant_damage.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                                        "italic": false
+                                    },
+                                    " ",
+                                    {
+                                        "translate": "potion.potency.3",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/instant_damage.json:30]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withAmplifier",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.instant_damage"
                                            },
                                            {
                                                "translate": "potion.potency.3"
                                            }
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/drinkable/instant_damage.json [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                                        "italic": false
+                                    },
+                                    " ",
+                                    {
+                                        "translate": "potion.potency.3",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/drinkable/instant_damage.json:146]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withAmplifier",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.instant_damage"
                                            },
                                            {
                                                "translate": "potion.potency.3"
                                            }
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/instant_health.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                                        "color": "red",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:02)",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/instant_health.json:42]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withAmplifier",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.instant_health"
                                            },
                                            {
                                                "translate": "potion.potency.3"
                                            }
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.nausea"
                                            },
                                            "0:02"
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/jump_boost.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                                        "color": "blue",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:22)",
+                                        "color": "blue",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/jump_boost.json:35]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.jump_boost"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:22"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/poison.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                                        "color": "red",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:8)",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/poison.json:35]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.poison"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:08"
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/regeneration.json [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                                        "color": "blue",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:4)",
+                                        "color": "blue",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/regeneration.json:53]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.regeneration"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:05"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.speed"
                                                    },
                                                    {
                                                        "translate": "potion.potency.2"
                                                    }
                                                ]
                                            },
                                            "0:04"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/speed.json [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                                    "with": [
+                                        "16",
+                                        {
+                                            "translate": "attribute.name.generic.attack_damage"
+                                        }
+                                    ],
+                                    "color": "blue",
+                                    "italic": false
+                                }
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/speed.json:83]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.speed"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:50"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.weakness"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "1:30"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                {
                                    "text": ""
                                },
                                {
                                    "translate": "potion.whenDrank",
                                    "color": "dark_purple",
                                    "italic": false
                                },
                                {
                                    "translate": "attribute.modifier.plus.1",
                                    "with": [
                                         "80",
                                         {
                                             "translate": "attribute.name.generic.movement_speed"
                                         }
                                     ],
                                     "color": "blue",
                                     "italic": false
                                },
                                {
                                    "translate": "attribute.modifier.take.0",
                                    "with": [
                                        "16",
                                        {
                                            "translate": "attribute.name.generic.attack_damage"
                                        }
                                    ],
                                    "color": "red",
                                    "italic": false
                                }
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/strength.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                                    "with": [
+                                        "12",
+                                        {
+                                            "translate": "attribute.name.generic.attack_damage"
+                                        }
+                                    ],
+                                    "color": "blue",
+                                    "italic": false
+                                }
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/lingering/strength.json:54]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.strength"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:22"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                {
                                    "text": ""
                                },
                                {
                                    "translate": "potion.whenDrank",
                                    "color": "dark_purple",
                                    "italic": false
                                },
                                {
                                    "translate": "attribute.modifier.plus.0",
                                    "with": [
                                        "12",
                                        {
                                            "translate": "attribute.name.generic.attack_damage"
                                        }
                                    ],
                                    "color": "blue",
                                    "italic": false
                                }
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/instant_damage.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                                        "italic": false
+                                    },
+                                    " ",
+                                    {
+                                        "translate": "potion.potency.3",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/instant_damage.json:30]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withAmplifier",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.instant_damage"
                                            },
                                            {
                                                "translate": "potion.potency.3"
                                            }
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/instant_health.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                                        "color": "red",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:08)",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/instant_health.json:42]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withAmplifier",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.instant_health"
                                            },
                                            {
                                                "translate": "potion.potency.3"
                                            }
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "effect.minecraft.nausea"
                                            },
                                            "0:08"
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/jump_boost.json [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+                                        "color": "blue",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (1:30)",
+                                        "color": "blue",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/jump_boost.json:35]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.jump_boost"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "1:30"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/poison.json [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+                                        "color": "red",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:32)",
+                                        "color": "red",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/poison.json:35]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.poison"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:32"
                                        ],
                                        "color": "red",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/regeneration.json [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+                                        "color": "blue",
+                                        "italic": false
+                                    },
+                                    {
+                                        "text": " (0:16)",
+                                        "color": "blue",
+                                        "italic": false
+                                    }
+                                ]
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/regeneration.json:53]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.regeneration"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "0:22"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.speed"
                                                    },
                                                    {
                                                        "translate": "potion.potency.2"
                                                    }
                                                ]
                                            },
                                            "0:16"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ]
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/speed.json [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+                                    "with": [
+                                        "16",
+                                        {
+                                            "translate": "attribute.name.generic.attack_damage"
+                                        }
+                                    ],
+                                    "color": "blue",
+                                    "italic": false
+                                }
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/speed.json:83]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.speed"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "3:20"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.weakness"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "6:00"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                {
                                    "text": ""
                                },
                                {
                                    "translate": "potion.whenDrank",
                                    "color": "dark_purple",
                                    "italic": false
                                },
                                {
                                    "translate": "attribute.modifier.plus.1",
                                    "with": [
                                         "80",
                                         {
                                             "translate": "attribute.name.generic.movement_speed"
                                         }
                                     ],
                                     "color": "blue",
                                     "italic": false
                                },
                                {
                                    "translate": "attribute.modifier.take.0",
                                    "with": [
                                        "16",
                                        {
                                            "translate": "attribute.name.generic.attack_damage"
                                        }
                                    ],
                                    "color": "red",
                                    "italic": false
                                }
                            ],
```

File: gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/strength.json [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+                                    "with": [
+                                        "12",
+                                        {
+                                            "translate": "attribute.name.generic.attack_damage"
+                                        }
+                                    ],
+                                    "color": "blue",
+                                    "italic": false
+                                }
+                            ],
```
Comment (BluePsychoRanger) [inline gm4_zauber_cauldrons/data/gm4_zauber_cauldrons/loot_tables/recipes/potions/splash/strength.json:54]:
```suggestion
                            "lore": [
                                [
                                    {
                                        "translate": "potion.withDuration",
                                        "with": [
                                            {
                                                "translate": "potion.withAmplifier",
                                                "with": [
                                                    {
                                                        "translate": "effect.minecraft.strength"
                                                    },
                                                    {
                                                        "translate": "potion.potency.3"
                                                    }
                                                ]
                                            },
                                            "1:30"
                                        ],
                                        "color": "blue",
                                        "italic": false
                                    }
                                ],
                                {
                                    "text": ""
                                },
                                {
                                    "translate": "potion.whenDrank",
                                    "color": "dark_purple",
                                    "italic": false
                                },
                                {
                                    "translate": "attribute.modifier.plus.0",
                                    "with": [
                                        "12",
                                        {
                                            "translate": "attribute.name.generic.attack_damage"
                                        }
                                    ],
                                    "color": "blue",
                                    "italic": false
                                }
                            ],
```

**Review by BluePsychoRanger — CHANGES_REQUESTED** [review]
I was looking over the loot tables, and I noticed you did the potion effects lore incorrectly (and a few typos). I've left suggestion comments to correctly display the potion effects correctly.

**Comment by Bloo-dev** [comment]
From my test wormholes seem to work. External testing would be appreciated.

**Comment by BluePsychoRanger** [comment]
Can you make the loot tables consistent with #662 please?

**Comment by Bloo-dev** [comment]
> Can you make the loot tables consistent with #663 please?

I don't think this is necessary. The loot table structure used in this module mirrors the function structure for increased readability.
Though I agree that `extra_items` is odd. I'll reorganize that before we merge this PR.

**Pull request merged**